src/page/page_cache.rs

```diff
@@ -359,7 +395,211 @@ impl<M> PageCache<LockFreeLog, M>
+        );
+
+        let mut recovery = None;
+
+        for (log_id, bytes) in self.log.iter_from(snapshot.max_lid) {
+            if log_id >= current_stable {
+                // don't need to go farther
+                break;
+            }
+            if let Ok(prepend) = deserialize::<LoggedUpdate<P>>(&*bytes) {
+                match prepend.update {
+                    Update::Append(partial_pages) => {
+                        for partial_page in partial_pages {
+                            let r = self.t.recover(&partial_page);
+                            if r.is_some() {
+                                recovery = r;
+                            }
+                        }
+
+                        let mut lids = snapshot.pt.get_mut(&prepend.pid).unwrap();
+                        lids.push(log_id);
+                    }
+                    Update::Compact(partial_pages) => {
+                        for partial_page in partial_pages {
+                            let r = self.t.recover(&partial_page);
+                            if r.is_some() {
+                                recovery = r;
+                            }
+                        }
+
+                        snapshot.pt.insert(prepend.pid, vec![log_id]);
+                    }
+                    Update::Del => {
+                        snapshot.pt.remove(&prepend.pid);
+                        snapshot.free.push(prepend.pid);
+                    }
+                    Update::Alloc => {
+                        snapshot.free.retain(|&pid| pid != prepend.pid);
+                        if prepend.pid > snapshot.max_pid {
+                            snapshot.max_pid = prepend.pid;
+                        }
+                    }
+                }
+            }
+        }
+        snapshot.free.sort();
+        snapshot.free.reverse();
+        snapshot.max_lid = current_stable;
+        snapshot.recovery = recovery;
+
+        let raw_bytes = serialize(&snapshot, Infinite).unwrap();
+        let bytes = if self.config().get_use_compression() {
+            compress(&*raw_bytes, 5).unwrap()
+        } else {
+            raw_bytes
+        };
+        let crc64: [u8; 8] = unsafe { std::mem::transmute(crc64::crc64(&*bytes)) };
+
+        self.last_snapshot.swap(snapshot, SeqCst);
+
+        let prefix = self.snapshot_prefix();
+        let path_1 = format!("{}.{}.in___motion", prefix, current_stable);
+        let path_2 = format!("{}.{}", prefix, current_stable);
+        let mut f = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(&path_1)
+            .unwrap();
+
+        // write the snapshot bytes, followed by a crc64 checksum at the end
+        f.write_all(&*bytes).unwrap();
+        f.write_all(&crc64).unwrap();
+        f.sync_all().unwrap();
+        drop(f);
+
+        if let Err(e) = std::fs::rename(path_1, &path_2) {
+            error!("failed to write snapshot: {}", e);
+        }
```

Comment:
Basic file write strategy:
1. open non-final path that signifies it's not finished with the name (`in___motion`)
1. write the snapshot and checksum
1. move (which is atomic) the file so that its name signifies that it's complete